src/txmempool.h

```diff
@@ -881,59 +875,79 @@ struct DisconnectedBlockTransactions {
     // to be refactored such that this assumption is no longer true (for
     // instance if there was some other way we cleaned up the mempool after a
     // reorg, besides draining this object).
-    ~DisconnectedBlockTransactions() { assert(queuedTx.empty()); }
-
-    indexed_disconnected_transactions queuedTx;
-    uint64_t cachedInnerUsage = 0;
+    ~DisconnectedBlockTransactions() {
+        assert(queuedTx.empty());
+        assert(iters_by_txid.empty());
+        assert(cachedInnerUsage == 0);
+    }
 
-    // Estimate the overhead of queuedTx to be 6 pointers + an allocation, as
-    // no exact formula for boost::multi_index_contained is implemented.
     size_t DynamicMemoryUsage() const {
-        return memusage::MallocUsage(sizeof(CTransactionRef) + 6 * sizeof(void*)) * queuedTx.size() + cachedInnerUsage;
+        return cachedInnerUsage + memusage::DynamicUsage(iters_by_txid) + memusage::DynamicUsage(queuedTx);
     }
 
-	/** Add transactions from the block, iterating through vtx in reverse order. Callers should call
+    /** Add transactions from the block, iterating through vtx in reverse order. Callers should call
      * this function for blocks in descending order by block height.
      * We assume that callers never pass multiple transactions with the same txid, otherwise things
      * can go very wrong in removeForBlock due to queuedTx containing an item without a
      * corresponding entry in iters_by_txid.
      */
     void AddTransactionsFromBlock(const std::vector<CTransactionRef>& vtx)
     {
+        iters_by_txid.reserve(iters_by_txid.size() + vtx.size());
         for (auto block_it = vtx.rbegin(); block_it != vtx.rend(); ++block_it) {
-            queuedTx.insert(*block_it);
-            cachedInnerUsage += RecursiveDynamicUsage(*block_it);
+            auto it = queuedTx.insert(queuedTx.end(), *block_it);
+            iters_by_txid.emplace((*block_it)->GetHash(), it);
```

Comment:
According to the comment: `We assume that callers never pass multiple transactions with the same txid`. 

IMO we should either commit to that and enforce it:
```c++
auto emplace_result = iters_by_txid.emplace((*block_it)->GetHash(), it);
assert(emplace_result.second);
```
or remove the restriction:
```c++
// Ignore duplicates
iters_by_txid.try_emplace((*block_it)->GetHash(), it);
```

Reply:
Agreed, I like the first option. Maybe @ismaelsadeeq might be interested in adding it to the followup?

Reply:
Added to #28530